cai_causal_graph/time_series_causal_graph.py

```diff
@@ -0,0 +1,880 @@
+
+
+class TimeSeriesNode(Node):
+    """
+    Time series node.
+
+    A node in a time series causal graph will have additional metadata and attributes that
+    gives the time information of the node together with the variable name.
+
+    The two additional metadata are:
+    - 'time_lag': the time difference with respect the reference time 0
+    - 'variable_name': the name of the variable (without the lag information).
+    """
+
+    def __init__(
+        self,
+        identifier: Optional[NodeLike] = None,
+        time_lag: Optional[int] = None,
+        variable_name: Optional[str] = None,
+        meta: Optional[Dict[str, Any]] = None,
+        variable_type: NodeVariableType = NodeVariableType.UNSPECIFIED,
+    ):
+        """
+        Initialize the time series node.
+
+        :param identifier: the identifier of the node. If the identifier is provided, the time_lag and
+            variable_name will be extracted from the identifier. Default is None.
+        :param time_lag: the time lag of the node. If `time_lag` is provided, then `variable_name` must be provided
+            to set the identifier. If both `time_lag` and `variable_name` are provided, the identifier must be None.
+            Default is None.
+        :param variable_name: the variable name of the node. If `variable_name` is provided, then `time_lag` must be
+            provided to set the identifier. If both `time_lag` and `variable_name` are provided, the identifier must
+            be None. Default is None.
+        :param meta: the metadata of the node. Default is None.
+        :param variable_type: the variable type of the node. Default is NodeVariableType.UNSPECIFIED.
+        """
+        if time_lag is not None and variable_name is not None:
+            assert identifier is None, 'If time_lag and variable_name are provided, identifier must be None.'
+            identifier = get_name_with_lag(variable_name, time_lag)
+        elif identifier is not None:
+            assert (
+                time_lag is None and variable_name is None
+            ), 'If identifier is provided, `time_lag` and `variable_name` must be None.'
+            identifier = Node.identifier_from(identifier)
+            variable_name, time_lag = get_variable_name_and_lag(identifier)
+        else:
+            raise ValueError(
+                'Either identifier or both time_lag and variable_name must be provided to initialize a node.'
+            )
+
+        # populate the metadata for each node
+        if meta is not None:
+            meta.update({'time_lag': time_lag, 'variable_name': variable_name})
+        else:
+            meta = {'time_lag': time_lag, 'variable_name': variable_name}
+
+        # populate the metadata for each node
+        super().__init__(identifier, meta, variable_type)
+
+    @property
+    def time_lag(self) -> int:
+        """Return the time lag of the node from the metadata."""
+        return self.meta.get('time_lag', 0)
```

Comment:
Should we return 0 or should we raise here? Seems like a requirement for a TimeSeriesnode that this information would exist. Was partly why I was saying we don't really need to store in meta anymore, it could be a protected class property `self._time_lag`

Reply:
I am still for storing in the meta. Can still raise here if `None` though.